### PR TITLE
Add query-frontend, query-scheduler and querier worker.

### PR DESCRIPTION
**Pull request by cyriltovena**
This enable the same read path as Mimir and Loki.

The query-frontend and query-scheduler are now new component of Phlare that ensure fair scheduling and queries acceleration in the future.

When using the All target (single binary) they are activated automatically. In distributed mode they are not mandatory but active by default.

The configuration rely on ring based service discovery using the mandatory memberlist configuration.

The code is mostly a copy/pasted from Mimir, but adapted  to work with connect-go. see connectgrpc package.

I'm planning to add runtime configs and tenant configs as a follow up.

### CODE REVIEW

File: pkg/phlare/modules.go [hunk 1]
```diff
@@ -217,6 +288,14 @@ func (f *Phlare) initServer() (services.Service, error) {
 	// see https://github.com/grafana/phlare/issues/231
 	f.Cfg.Server.DoNotAddDefaultHTTPMiddleware = true
 
+	// Disable timeouts if query-scheduler is enabled since it requires long streaming RPCs.
+	if f.isModuleActive(QueryScheduler) {
+		// todo: however we should have API timeout for queries when running a single binary.
```
Comment (simonswine) [inline pkg/phlare/modules.go:293]: Probably worth creating an GitHub issue for that

Comment (simonswine) [inline pkg/phlare/modules.go:293]: Probably worth creating an GitHub issue for that.

File: pkg/phlare/modules.go [hunk 2]
```diff
@@ -217,6 +288,14 @@ func (f *Phlare) initServer() (services.Service, error) {
...
+		f.Cfg.Server.HTTPServerIdleTimeout = 0
+		f.Cfg.Server.HTTPServerWriteTimeout = 0
+		f.Cfg.Server.HTTPServerReadTimeout = 0
```
Comment (simonswine) [inline pkg/phlare/modules.go:296]: `0` is quite long, maybe we can find some middle ground. How big is the disruption when a timeout happens? 

I am thinking of scenarios like:
- if the connection is really dead on a e.g. TCP level?
- how do certificates ever getting renewed?

File: pkg/querier/stats/stats.proto [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+syntax = "proto3";
+
+package stats;
+
+option go_package = "stats";
+
+message Stats {
+  // The sum of all wall time spent in the querier to execute the query.
+  int64 wall_time = 1;
```
Comment (simonswine) [inline pkg/querier/stats/stats.proto:14]: Maybe wall_time_seconds?

File: pkg/querier/stats/stats.proto [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  // The number of series fetched for the query
+  uint64 fetched_series_count = 2;
+  // The number of bytes of the chunks fetched for the query
+  uint64 fetched_chunk_bytes = 3;
+  // The number of chunks fetched for the query
+  uint64 fetched_chunks_count = 4;
+  // The number of sharded queries executed. 0 if sharding is disabled or the query can't be sharded.
+  uint32 sharded_queries = 5;
```
Comment (simonswine) [inline pkg/querier/stats/stats.proto:22]:
```suggestion
  uint32 sharded_queries_count = 5;
```

**Review by simonswine — COMMENTED** [review]
I have not looked at all code, but that looks quite good sofar, a couple of nits, which I have noticed along the way.

File: pkg/querier/stats/stats.proto (hunk 3 above)
Comment (cyriltovena) [reply pkg/querier/stats/stats.proto:14]: Those are from mimir, and not fully sure they will stay as it. I kept them for simplicity.

File: pkg/phlare/modules.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/phlare/modules.go:296]: For context the frontend and querier will enter this loop on reaching timeout:

```
backoffConfig := backoff.Config{
		MinBackoff: 250 * time.Millisecond,
		MaxBackoff: 2 * time.Second,
	}
	backoff := backoff.New(ctx, backoffConfig)
	for backoff.Ongoing() {
		if !attemptLoop() {
			backoff.Wait()
		} else {
			backoff.Reset()
		}
	}
```
So 250ms before retrying.

Comment (cyriltovena) [reply pkg/phlare/modules.go:296]: This is how it looks like 

```
level=debug caller=gokit.go:65 caller=http.go:183 traceID=4a5eaf1a36f6d586 msg="POST /schedulerpb.SchedulerForFrontend/FrontendLoop (200) 29.996462956s"
level=debug caller=gokit.go:65 caller=http.go:183 traceID=1e01b3ce743005ec msg="POST /schedulerpb.SchedulerForFrontend/FrontendLoop (200) 30.001080655s"
level=debug caller=gokit.go:65 caller=http.go:183 traceID=1b5d7beb9d0fdb9d msg="POST /schedulerpb.SchedulerForFrontend/FrontendLoop (200) 29.996441541s"
level=error caller=scheduler_processor.go:128 ts=2023-02-06T10:52:44.898988Z component=querier-worker msg="error processing requests from scheduler" err="rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: INTERNAL_ERROR" addr=192.168.1.194:4100
level=error caller=scheduler_processor.go:128 ts=2023-02-06T10:52:44.899003Z component=querier-worker msg="error processing requests from scheduler" err="rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: INTERNAL_ERROR" addr=192.168.1.194:4100
level=error caller=scheduler_processor.go:128 ts=2023-02-06T10:52:44.90318Z component=querier-worker msg="error processing requests from scheduler" err="rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: INTERNAL_ERROR" addr=192.168.1.194:4100
level=error caller=scheduler_processor.go:128 ts=2023-02-06T10:52:44.903196Z component=querier-worker msg="error processing requests from scheduler" err="rpc error: code = Internal desc = stream terminated by RST_STREAM with error code: INTERNAL_ERROR" addr=192.168.1.194:4100
```

**Review by simonswine — APPROVED** [review]
LGTM. I have tried this all on a local environment and I think this is ready to go

**Pull request merged**